### DISCUSSION TITLE
Power BI DBX Connectivity - Simba Spark Driver Installation

### DISCUSSION THREAD

**Discussion by Koushikrishnan** [TabularEditor/TabularEditor3#1293] [🙋♀️TE3 Q&A]
Hello!

We had a question regarding a driver installation for Power BI DBX connectivity.

To give a quick background of our ways of development - We are an org that uses Tabular Editor solely for our development of Power BI Datasets. We use both TE3 and TE2 portable. Recently as part of a architectural initiative, we have been migrating all our data into Databricks Unity Catalog tables. The Power BI datasets that consume these data are again developed via TE - we have templates with Power Query for the dataset table partitions and use that accordingly to pull in data into the datasets. This has been working well without issues. Also for moving our changes to our development environments to Production workspaces, we have been using Github actions through Power BI deployment pipelines which too has been working well.

A few days back we encountered a Memory related error on Power BI which caused one of our most important datasets to fail and affect our business productivity. We raised an incident with Microsoft over this who had been looking into identifying the root cause behind this. Based on their most recent response last week, they identified that this could be due to a missing driver installation. Please find their response below:

"

_We have a validation from our Product Group Team that even if you are connecting and updating the Semantic Model (Power BI Dataset) using Tabular Editor by establishing connection through XMLA endpoint. For all Databricks connections with Power BI the Simba Spark Driver is used. 

As mentioned Power BI integration with Tabular Editor requires the installation of Power BI Desktop, and the Power BI Desktop releases include Simba ODBC driver pre-installed. 
Tabular Editor’s documentation: [Power BI Desktop Integration | Tabular Editor Documentation](https://url.uk.m.mimecastprotect.com/s/lTnrCk7LQh35Mmgnh29yf8?domain=docs.tabulareditor.com)

The Root Cause Analysis is correct for the failed refreshes experienced Tuesday 4/23 at ~1:30am and ~3:11am EST it was because the Simba drivers encountered difficulties closing an active connection in some instances with Power BI ending the operation after exceeding the memory available displaying that in the error message.

The initial workaround involved two steps to work as mitigation:
1.	A feature request has been made to the partner ecosystem Simba team to address the driver issue. 
2.	Databricks implemented a 15-minute timeout in these scenarios, as opposed to the current 1-hour timeout.

The recommended action step is update the ODBC driver to the latest version, to ensure the Simba driver includes latest fixes available. [Download and install the Databricks ODBC Driver - Azure Databricks | Microsoft Learn](https://url.uk.m.mimecastprotect.com/s/gN07Cl7gQhqXBgLoc9I4L7?domain=learn.microsoft.com) You can implement this with an on-premises data gateway._

"

We have explained to Microsoft regarding this and we are intending to deploy the suggested Simba driver on our development machines as well. We also checked with them on which other places this driver needs to be installed to ensure there is no issue with moving the data to Production via the pipeline. Microsoft suggested we contact Tabular Editor regarding this and help us identify which other areas this would be required to be installed. 

Could you please go through the information provided and let us know what other steps we need to take in consideration to ensure the driver installation is not an issue anymore? Any documentation, suggestion on areas (actions, pipelines, etc.) where the driver needs to be installed or any other steps to tackle this would be great. Looking forward to your response!

Thank you.
Venkat Narayanan.

**Comment by Koushikrishnan** [comment]
Hello., just following up to see if we can get an update on this. @otykier

**Reply by otykier** [reply]
Hi @Koushikrishnan 
I'm very sorry, but I'm not sure what your expectation is here. Tabular Editor only establishes a connection to Databricks while you're using the **Table Import Wizard** or the **Update Table Schema** feature. It uses this connection only to bring table metadata (i.e. column names and data types) into the model. This is completely unrelated to any memory issues encountered in Power BI. Perhaps I'm misunderstanding the question?

**Reply by Koushikrishnan** [reply]
@otykier Thanks for the response. Your answer is what we have been thinking as well but Microsoft gave us a response as mentioned in my original post to link the memory issue to the ODBC driver. We use the Github pipeline which uses tabular editor portable to run the deployment to Prod, so would just like to know if we would need to install the driver they suggested anywhere else apart from the Development machine.

**Reply by Koushikrishnan** [reply]
FYI one of our engineers got this error when working on TE3 on his dev machine yesterday which should be resolved after the Driver MS suggested was installed. So we would just like to know if that installation on the Dev machine is only driver installation we would need to do.

![image](https://github.com/TabularEditor/TabularEditor3/assets/87026967/d0ff25cb-2e43-4029-9628-725da9cd4726)

**Reply by otykier** [reply]
The ODBC driver is not needed on the machine used by the GitHub pipeline to perform the deployment. The driver is only needed on dev machines where users want to import/update tables from Databricks. However, it doesn't matter which version of the driver is used on the dev machine, as the table metadata that is imported, will be exactly the same in any case, and so this couldn't possibly impact the memory usage of Power BI (which uses its own set of Simba drivers to connect, completely unrelated to what you have installed on your dev machine).

**Reply by Koushikrishnan** [reply]
Got it. I think we should be good then. I will get back in case of any other questions but for now this is what I wanted to confirm. Thank you very much @otykier!
